File: src/algorithms/midpointCircle.ts
import { CircleStepState, CircleSymmetricPoints, PixelCoord, Point2D } from '../types';

export function getSymmetricPoints(xc: number, yc: number, x: number, y: number): CircleSymmetricPoints {
  return {
    octant1: { x: xc + x, y: yc + y },
    octant2: { x: xc - x, y: yc + y },
    octant3: { x: xc + x, y: yc - y },
    octant4: { x: xc - x, y: yc - y },
    octant5: { x: xc + y, y: yc + x },
    octant6: { x: xc - y, y: yc + x },
    octant7: { x: xc + y, y: yc - x },
    octant8: { x: xc - y, y: yc - x }
  };
}

export function generateMidpointCircleSteps(xc: number, yc: number, radius: number): CircleStepState[] {
  const steps: CircleStepState[] = [];
  const r = Math.round(Math.abs(radius));

  if (r === 0) {
    const symm = getSymmetricPoints(xc, yc, 0, 0);
    return [{
      step: 0,
      totalSteps: 0,
      x: 0,
      y: 0,
      pk: 0,
      center: { x: xc, y: yc },
      radius: 0,
      symmetricPoints: symm,
      explanation: `Radius is 0. Single center pixel plotted at (${xc}, ${yc}).`,
      historyPixels: [{ x: xc, y: yc }]
    }];
  }

  let x = 0;
  let y = r;
  let pk = 1 - r;

  // Pre-calculate step count for accurate timeline
  let simX = 0;
  let simY = r;
  let simP = 1 - r;
  let count = 0;
  while (simX <= simY) {
    count++;
    if (simP < 0) {
      simP += 2 * simX + 3;
    } else {
      simP += 2 * (simX - simY) + 5;
      simY--;
    }
    simX++;
  }
  const totalSteps = count - 1;

  const historyMap = new Map<string, PixelCoord>();

  const addSymmetricToHistory = (symm: CircleSymmetricPoints) => {
    const pts: Point2D[] = [
      symm.octant1, symm.octant2, symm.octant3, symm.octant4,
      symm.octant5, symm.octant6, symm.octant7, symm.octant8
    ];
    pts.forEach((p, idx) => {
      const key = `${p.x},${p.y}`;
      if (!historyMap.has(key)) {
        historyMap.set(key, { x: p.x, y: p.y, symmetricGroup: idx + 1 });
      }
    });
  };

  let stepIndex = 0;

  while (x <= y) {
    const symm = getSymmetricPoints(xc, yc, x, y);
    addSymmetricToHistory(symm);

    let nextPk = pk;
    let nextY = y;
    let expl = '';

    if (stepIndex === 0) {
      expl = `Initial point (x=0, y=${r}) relative to center (${xc}, ${yc}). Initial decision parameter P₀ = 1 - R = 1 - ${r} = ${pk}. 8 symmetric pixels generated on the grid.`;
    }

    if (pk < 0) {
      nextPk = pk + 2 * x + 3;
      nextY = y;
      if (stepIndex > 0) {
        expl = `Step ${stepIndex}: Since P = ${pk} < 0, midpoint lies inside circle. Choose E (x+1, y) = (${x + 1}, ${y}). Next P = P + 2x + 3 = ${pk} + 2(${x}) + 3 = ${nextPk}.`;
      }
    } else {
      nextPk = pk + 2 * (x - y) + 5;
      nextY = y - 1;
      if (stepIndex > 0) {
        expl = `Step ${stepIndex}: Since P = ${pk} ≥ 0, midpoint lies outside/on circle. Choose SE (x+1, y-1) = (${x + 1}, ${y - 1}). Next P = P + 2(x - y) + 5 = ${pk} + 2(${x - y}) + 5 = ${nextPk}.`;
      }
    }

    steps.push({
      step: stepIndex,
      totalSteps,
      x,
      y,
      pk,
      pkNext: nextPk,
      center: { x: xc, y: yc },
      radius: r,
      symmetricPoints: symm,
      explanation: expl,
      historyPixels: Array.from(historyMap.values())
    });

    pk = nextPk;
    x++;
    y = nextY;
    stepIndex++;
  }

  return steps;
}

File: python_project/algorithms/midpoint_circle.py
"""
Midpoint Circle Drawing Algorithm with 8-way symmetry.
Calculates coordinates for one octant and reflects across all 8 octants.
"""

class MidpointCircle:
    def __init__(self, xc: int, yc: int, radius: int):
        self.xc = int(xc)
        self.yc = int(yc)
        self.radius = int(radius)

    def get_symmetric_points(self, x: int, y: int):
        return [
            (self.xc + x, self.yc + y),
            (self.xc - x, self.yc + y),
            (self.xc + x, self.yc - y),
            (self.xc - x, self.yc - y),
            (self.xc + y, self.yc + x),
            (self.xc - y, self.yc + x),
            (self.xc + y, self.yc - x),
            (self.xc - y, self.yc - x)
        ]

    def generate_steps(self):
        r = abs(self.radius)
        if r == 0:
            return [{
                "step": 0,
                "total_steps": 0,
                "x": 0,
                "y": 0,
                "pk": 0,
                "symmetric_points": [(self.xc, self.yc)],
                "explanation": f"Radius is 0. Plotted center ({self.xc}, {self.yc}).",
                "pixels": [(self.xc, self.yc)]
            }]

        x = 0
        y = r
        pk = 1 - r

        # Count total steps
        sim_x, sim_y, sim_p = 0, r, 1 - r
        total_steps = 0
        while sim_x <= sim_y:
            total_steps += 1
            if sim_p < 0:
                sim_p += 2 * sim_x + 3
            else:
                sim_p += 2 * (sim_x - sim_y) + 5
                sim_y -= 1
            sim_x += 1
        total_steps -= 1

        step_records = []
        accumulated_pixels = set()
        step_idx = 0

        while x <= y:
            symm = self.get_symmetric_points(x, y)
            for pt in symm:
                accumulated_pixels.add(pt)

            if step_idx == 0:
                expl = f"Start at (0, R)=(0, {r}). Initial P0 = 1 - R = 1 - {r} = {pk}. Generated 8 symmetric points."
            elif pk < 0:
                next_pk = pk + 2 * x + 3
                expl = f"Step {step_idx}: P = {pk} < 0 (inside). Choose E (x+1, y)=({x+1}, {y}). Next P = P + 2x + 3 = {next_pk}."
            else:
                next_pk = pk + 2 * (x - y) + 5
                expl = f"Step {step_idx}: P = {pk} >= 0 (outside/on). Choose SE (x+1, y-1)=({x+1}, {y-1}). Next P = P + 2(x-y) + 5 = {next_pk}."

            if pk < 0:
                next_pk = pk + 2 * x + 3
                next_y = y
            else:
                next_pk = pk + 2 * (x - y) + 5
                next_y = y - 1

            step_records.append({
                "step": step_idx,
                "total_steps": total_steps,
                "x": x,
                "y": y,
                "pk": pk,
                "pk_next": next_pk,
                "symmetric_points": symm,
                "explanation": expl,
                "pixels": list(accumulated_pixels)
            })

            pk = next_pk
            x += 1
            y = next_y
            step_idx += 1

        return step_records
